src/test/java/com/alicetechnologies/planner/CriticalPathEngineTest.java

```diff
@@ -0,0 +1,75 @@
+package com.alicetechnologies.planner;
+
+import com.alicetechnologies.planner.task.dto.Task;
+import com.alicetechnologies.planner.task.dto.TaskEvaluated;
+import com.alicetechnologies.planner.task.TaskRepository;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class CriticalPathEngineTest {
+
+    /**
+     * C -> B -> A
+     * D -> E
+     */
+    final List<Task> sourceTasks = List.of(
+        task("B", "build walls", 20, "A"),
+        task("A", "lay foundation", 5),
+        task("C", "build roof", 4, "B"),
+        task("D", "eat lunch", 3, "E"),
+        task("E", "drink coffee", 2)
+    );
+
+    @Test
+    void tinyDataset() {
+        CriticalPathEngine engine = new CriticalPathEngine(sourceTasks);
+
+        final String taskOrder = engine.getTasks().stream()
+            .sorted(Comparator.comparingInt(TaskEvaluated::getEarlyStart))
+            .map(TaskEvaluated::getTaskCode)
+            .collect(Collectors.joining(", "));
+
+        final String taskStarts = engine.getTasks().stream()
+            .sorted(Comparator.comparingInt(TaskEvaluated::getEarlyStart))
+            .map(task -> String.format("%s=%d", task.getTaskCode(), task.getEarlyStart()))
+            .collect(Collectors.joining(", "));
+
+        assertAll(
+            () -> assertEquals(29, engine.getMaxCost()),
+            () -> assertEquals("C, D, E, B, A", taskOrder), // this doesn't look right, I'd expect the opposite order
```

Comment:
A bit confused here. I would expect that `C -> B -> A` dependencies relationship means that `A` needs to be done first, then `B`, then `C`.

But the algorithm seems to treat `dependencies` as the tasks that are blocked, not as the tasks that are blocking.
I'm not familiar with this area, so please advise.

Reply:
I need to check the code to answer correctly. But we define dependencies list as all the tasks that need to be completed before this task can start. So in the following input

```
[
    { "taskCode" : "A", "duration" : 8, "dependencies" : [ ], ...},
    { "taskCode" : "B", "duration" : 8, "dependencies" : [ "B" ], ...},
    { "taskCode" : "C", "duration" : 8, "dependencies" : [ "C" ], ...},
    { "taskCode" : "D", "duration" : 8, "dependencies" : [ "A" ], ...}
]
```

The final sequence should be something like:

```
[
    { "taskCode" : "A", "start" : 0, "finish": 8, ...},
    { "taskCode" : "B", "start" : 8, "finish": 16, ...},
    { "taskCode" : "C", "start" : 16, "finish": 24, ...},
    { "taskCode" : "D", "start" : 8, "finish": 16, ...}
]
```

So in your case I would expect the order of work to be `A, B, C` but I'm not sure if I'm seeing the final order or some intermediate part of the process at this point.

I need to check your implementation and the respective test to make sure that this is the case.

Reply:
Thank you for clarification! What you are seeing in the test is the final order. Based on your description, the order should have been `A -> B -> C`, and so the current result is not correct and I need to investigate it more.

Reply:
The issue has been now fixed, the order of execution now matches what is expected, as per discussion.